### ISSUE TITLE
Implementation of Thresholding of Gradients -NeuRD

### ISSUE THREAD

**Issue by chiggy2402** [deepmind/open_spiel#316]
I was working with `open_spiel/open_spiel/python/algorithms/neurd.py` and was trying implement thresholding part in my version of Gradient Update. However it seems like the Remark in the NRD paper about the gradient clipping and it's implementation in OpenSpiel don’t quite match up.  In particular the latter just checks whether the logit is already too large and then zeros out the gradient if it is, where the former does a more complicated check (which maybe requires an additional forward pass? since it used the `θ' = θ + η∇z(θ) and calculates z(θ') ∈ [−β, β)` ) about whether the updated parameters would have resulted in a logic * regret that is too large.  

There seem to be several differences between the two – whether to check for threshold before / after taking the update into account and whether z = logit or z = logit*regret, and why the paper describes this as bound the logit-gap when each logit seems to be clipped in isolation.  It also isn’t clear whether the code implementation matches what you described about this for the paper.

**Comment by lanctot** [comment]
HI @chiggy2402, the implementation used for most of the experiments in the paper was not open-sourced. I believe this one was used for the convergence over time comparison to SPG in Leduc poker.

@dmorrill10, can you say more?

Btw you might also be interested in this discussion among the main authors: https://github.com/deepmind/open_spiel/issues/148

**Comment by dmorrill10** [comment]
Thanks for the question, @chiggy2402, and thanks for the ping @lanctot.

The remark at the end of page 5 and the beginning of page 6 on [the most recent version of the paper on arXiv](https://arxiv.org/pdf/1906.00190.pdf) states that the gradient is thresholded before the update as the code describes (clears the update to zero if the unthresholded gradient update would make the magnitude of the logits too large). This discussion sounds a bit familiar, I think in an early version of the paper the thresholding was described like this and it was changed after we noticed that it didn't match what we actually ran. IIRC, I copied the thresholding code exactly from the private implementation of NeuRD, so what is in `open_spiel/open_spiel/python/algorithms/neurd.py` should be what was done in this respect for the function approximation experiments.

Does that answer your questions? Please let me know if something is still unclear.

**Comment by lanctot** [comment]
Quick comment: I had a chat about this with @perolat and he said it was consistent with what `neurd.py` does. I've asked him to elaborate.

**Comment by perolat** [comment]
Hi all,

`
def apply_force_with_threshold(decision_outputs, force, threshold):
  can_decrease = tf.cast(tf.greater(decision_outputs, -threshold), tf.float32)
  can_increase = tf.cast(tf.less(decision_outputs, threshold), tf.float32)
  force_negative = tf.minimum(force, 0.0)
  force_positive = tf.maximum(force, 0.0)
  clipped_force = can_decrease * force_negative + can_increase * force_positive
  return decision_outputs * tf.stop_gradient(clipped_force)

def loss_nerd_update(logits, legal_actions, policy, q_values, threshold):
  """Creates the neurd loss of the policy.
  Args:
    logits: the policy logits [B, T, Na].
    legal_actions: the legal actions
    policy: the policy on the legal action for the observation [B, T, Na].
    q_values: the q-values [B, T, Na].
    threshold: a logit threshold
  Returns:
    The loss
  """
  adv_pi = tf.stop_gradient(q_values) - tf.reduce_sum(
      policy * tf.stop_gradient(q_values), axis=-1, keep_dims=True)
  logits = logits - tf.reduce_mean(logits * legal_actions, -1, keep_dims=True)
  policy_gradient_loss = -tf.reduce_sum(
      legal_actions * apply_force_with_threshold(logits, adv_pi, threshold),
      axis=-1)
  loss_policy = tf.reduce_sum(policy_gradient_loss)
  return loss_policy
`

here is a cleaned up version of the code as we used it for the experiments. To the best of my understanding it is what is implemented in the function `thresholded` in `open_spiel/open_spiel/python/algorithms/neurd.py`.

Hope this is helpful!

**Comment by chiggy2402** [comment]
Thanks for your quick replies @dmorrill10 @lanctot @perolat . I understand the code implementation. However, I am still not sure if it matches the remark in the recent version of paper. 

1. Do you consider` z = y ` (as seen in the algorithm 1, line 10) or `z = y * (q - v) ` ( as seen in the remark's theta update equation)?

2. Is the thresholding applied before the update ` z(θ) ∈ [−β, β)` (as seen in the code) or  after the update `θ' = θ + η∇z(θ),   z(θ') ∈ [−β, β)` (as seen in the remark)

Please let me know if I am not reading something correctly.

**Comment by dmorrill10** [comment]
The pseudocode update does not contain the threshold operation. Line 10 is swapped out for the second line of math in the remark at the start of page 6. The parameters are adjusted according to the `hat{nabla}` function, which is the thresholded gradient function. So the dot product of the logits and the regret becomes the `z(theta)` parameter of `hat{nabla}`. `hat{nabla}` computes the gradient of this dot product and returns the clipped gradient, which is added to the parameters, `theta`.

**Comment by chiggy2402** [comment]
Ok. So I understand that you consider` z = logits * regret `which is used to compute the clipped gradient and then finally you update the parameters. However, can you explain why the paper says `z(θ + η∇z(θ)) ∈ [−β, β)` ? Shouldn't it be `z(θ) ∈ [−β, β)`?

**Comment by dmorrill10** [comment]
Oh, thanks, I think you're right that there is a mistake there but it's a little different from your suggestion. `hat{∇}` needs to take the logits and regrets as separate arguments and return the gradient of the logits dot regrets times the indicator that the *logits* are in `[−β, β)`. So `hat{∇}(y, r, η, β) = η∇<y, f(r, y, β)>` where `f(r, y, β) = r I{y ∈ [−β, β)}`. @perolat, can you crosscheck the math and code, or ask Daniel or Shayegan to take a look?

**Comment by chiggy2402** [comment]
Thanks for the clarification.

Can you explain as to why you have z(θ + η∇z(θ)) ∈ [−β, β) inside the indicator? Shouldn't it be z(θ) ∈ [−β, β)?

**Comment by lanctot** [comment]
@chiggy2402 did you ever get a sample-based impl of NeuRD fully working? Would you be interested in contributing it via a pull request?
